server/principal.go

```diff
@@ -49,30 +49,13 @@ func (s *Server) registerPrincipalRoutes(g *echo.Group) {
 
 	g.GET("/principal", func(c echo.Context) error {
 		ctx := context.Background()
-		principalList, err := s.PrincipalService.FindPrincipalList(ctx)
+		principalList, err := s.PrincipalService.FindList(ctx)
 		if err != nil {
 			return echo.NewHTTPError(http.StatusInternalServerError, "Failed to fetch principal list").SetInternal(err)
 		}
 
-		filteredList := []*api.Principal{}
-		for _, principal := range principalList {
-			if err := s.composePrincipalRole(ctx, principal); err != nil {
-				// Normally this should not happen since we create the member together with the principal
```

Comment:
This is a hack, if there's any chance that we break the assumption, we should fix it.

Reply:
Because we can't enforce it on the database schema level, thus this could happen. The if block is to provide a more explicit error message when that happens.

Reply:
There's another issue: if we consider this possible, then any call to `ComposePrincipalByID()` could potentially result in a `Principal` without a corresponding role.

If it's really an issue, we should fix it as a whole.

Reply:
yeah, returning an error something like "the principal has not been assigned a role" is the way to go.

Reply:
LGTM. Let's be clear, if we encounter a `Principal`without a corresponding `Role`, do we:

1. log an error message and continue

2. stop the process and return an error

WDYT? @tianzhou

Reply:
I prefer 2.

Reply:
Oh, I found another problem down there...

We are returning `nil, nil` when nothing found in the database, but checking for `common.ErrorCode(err) == common.NotFound` here.

Maybe we should check all the error returning when refactoring, and return `nil, wrapped common.NotFound` instead?

@tianzhou

Reply:
SGTM